File: examples/governance/model_registry_domain/lib/model_registry_domain/ai_model/commands/register_model.rb
module ModelRegistryDomain
  class AiModel
    module Commands
      class RegisterModel
        include Hecks::Command
        emits "RegisteredModel"

        attr_reader :name
        attr_reader :version
        attr_reader :provider_id
        attr_reader :description

        def initialize(
          name: nil,
          version: nil,
          provider_id: nil,
          description: nil
        )
          @name = name
          @version = version
          @provider_id = provider_id
          @description = description
        end

        def call
          AiModel.new(
            name: name,
            version: version,
            provider_id: provider_id,
            description: description,
            registered_at: Time.now.to_s,
            status: "draft"
          )
        end
      end
    end
  end
end

File: examples/governance_static_go/domain/register_model.go
package domain

import (
	"time"
)

type RegisterModel struct {
	Name string `json:"name"`
	Version string `json:"version"`
	ProviderId string `json:"provider_id"`
	Description string `json:"description"`
}

func (c RegisterModel) CommandName() string { return "RegisterModel" }

func (c RegisterModel) Execute(repo AiModelRepository) (*AiModel, *RegisteredModel, error) {
	agg := NewAiModel(c.Name, c.Version, c.ProviderId, c.Description, "", time.Time{}, "", "", nil, nil, "")
	agg.Status = "draft"
	if err := agg.Validate(); err != nil {
		return nil, nil, err
	}
	if err := repo.Save(agg); err != nil {
		return nil, nil, err
	}
	event := RegisteredModel{
		AggregateID: agg.ID,
		Name: c.Name,
		Version: c.Version,
		ProviderId: c.ProviderId,
		Description: c.Description,
		OccurredAt: time.Now(),
	}
	return agg, &event, nil
}
